File: HMCLCore/src/main/java/org/jackhuang/hmcl/auth/yggdrasil/GameProfile.java
/*
 * Hello Minecraft! Launcher
 * Copyright (C) 2020  huangyuhui <huanghongxun2008@126.com> and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package org.jackhuang.hmcl.auth.yggdrasil;

import java.util.Objects;
import java.util.UUID;

import org.jackhuang.hmcl.util.Immutable;
import org.jackhuang.hmcl.util.gson.UUIDTypeAdapter;
import org.jackhuang.hmcl.util.gson.Validation;

import com.google.gson.JsonParseException;
import com.google.gson.annotations.JsonAdapter;

/**
 * @author huangyuhui
 */
@Immutable
public class GameProfile implements Validation {

    @JsonAdapter(UUIDTypeAdapter.class)
    private final UUID id;

    private final String name;

    public GameProfile(UUID id, String name) {
        this.id = Objects.requireNonNull(id);
        this.name = Objects.requireNonNull(name);
    }

    public UUID getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    @Override
    public void validate() throws JsonParseException {
        Validation.requireNonNull(id, "Game profile id cannot be null");
        Validation.requireNonNull(name, "Game profile name cannot be null");
    }
}

File: HMCLCore/src/main/kotlin/org/jackhuang/hmcl/auth/yggdrasil/GameProfile.kt
/*
 * Hello Minecraft! Launcher.
 * Copyright (C) 2017  huangyuhui <huanghongxun2008@126.com>
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see {http://www.gnu.org/licenses/}.
 */
package org.jackhuang.hmcl.auth.yggdrasil

import com.google.gson.*
import java.lang.reflect.Type
import java.util.*

data class GameProfile(
        val id: UUID? = null,
        val name: String? = null,
        val properties: PropertyMap = PropertyMap(),
        val legacy: Boolean = false
) {
    companion object GameProfileSerializer: JsonSerializer<GameProfile>, JsonDeserializer<GameProfile> {
        override fun serialize(src: GameProfile, typeOfSrc: Type?, context: JsonSerializationContext): JsonElement {
            val result = JsonObject()
            if (src.id != null)
                result.add("id", context.serialize(src.id))
            if (src.name != null)
                result.addProperty("name", src.name)
            return result
        }

        override fun deserialize(json: JsonElement?, typeOfT: Type?, context: JsonDeserializationContext): GameProfile {
            if (json !is JsonObject)
                throw JsonParseException("The json element is not a JsonObject.")
            val id = if (json.has("id")) context.deserialize<UUID>(json.get("id"), UUID::class.java) else null
            val name = if (json.has("name")) json.getAsJsonPrimitive("name").asString else null
            return GameProfile(id, name)
        }

    }
}